### PR TITLE
completed assessments

### PR DESCRIPTION
**Pull request by SOMarrah**
# Submitting Week 5 Assessment

### ✅ Completed Files
Mark the following as confirmation that the files in this repository are complete.
- [x] JavaScript Coding Challenges
- [x] Ruby Coding Challenges
- [x] Interview Questions

### 📝 Feedback
If you would like specific feedback on a particular problem please leave details here:
This week was a lot of fun.

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -15,42 +15,252 @@
...
 
-const secretCodeWord1 = "Lackadaisical"
-// Expected output: "L4ck4d41s1c4l"
-const secretCodeWord2 = "Gobbledygook"
-// Expected output: "G0bbl3dyg00k"
-const secretCodeWord3 = "Eccentric"
-// Expected output: "3cc3ntr1c"
+
+
+describe("l33t", () => {
```
Comment (SunkissedQueen) [inline code-challenges.test.js:20]: like that function name

File: code-challenges.test.js [hunk 2]
```diff
@@ -15,42 +15,252 @@
...
+  //use a conditional to make a decision
+    //if current letter is one of the specific letters replace current letter with specific letter
+    //else leave the current letter
+ //-maybe break the string into an array, compare each letter to the specific letters, push those values into a new array, then return that new array combined into a string?
+//tools and research: .split() splits a string into an array of substrings - returns the new array. Does not change the original string. .toString() turns an array into a string with commas. .join() turns an array into a string - does not change the original array. regex - a pattern to match character combinations in strings.
+  //maybe can replace the string directly with regex like in ruby?
+    //.replace() returns a new string with some/all matches to a pattern. 
+
+//Thinking this through fully before typing any code helped here. Researching regex was super useful. To refactor maybe I can try combining all of the regex into one? will sperators work like in Ruby? How will I identify to replace 'a' with '4' and 'e' with '3' inside on a continuous regex?
+const l33t = (string) => {
```
Comment (SunkissedQueen) [inline code-challenges.test.js:73]: looks great

File: code-challenges.test.js [hunk 3]
```diff
@@ -15,42 +15,252 @@
...
+  //use a conditional to make a decision
+    //if current letter is one of the specific letters replace current letter with specific letter
+    //else leave the current letter
+ //-maybe break the string into an array, compare each letter to the specific letters, push those values into a new array, then return that new array combined into a string?
+//tools and research: .split() splits a string into an array of substrings - returns the new array. Does not change the original string. .toString() turns an array into a string with commas. .join() turns an array into a string - does not change the original array. regex - a pattern to match character combinations in strings.
+  //maybe can replace the string directly with regex like in ruby?
+    //.replace() returns a new string with some/all matches to a pattern. 
+
+//Thinking this through fully before typing any code helped here. Researching regex was super useful. To refactor maybe I can try combining all of the regex into one? will sperators work like in Ruby? How will I identify to replace 'a' with '4' and 'e' with '3' inside on a continuous regex?
```
Comment (SunkissedQueen) [inline code-challenges.test.js:72]: All great questions. Looking forward to your research.

File: code-challenges.test.js [hunk 4]
```diff
@@ -15,42 +15,252 @@
...
+        // Expected output: ["Apple", "Banana", "Orange"]
+        const arrayOfWords2 = ["Mango", "Cherry", "Apricot", "Blueberry", "Peach"]
+        const letterE = "e"
+        // Expected output: ["Cherry", "Blueberry", "Peach"]
+      expect(arrayMatch(arrayOfWords1,letterA)).toEqual(["Apple", "Banana", "Orange"])
+      expect(arrayMatch(arrayOfWords2,letterE)).toEqual(["Cherry", "Blueberry", "Peach"])
+    })
+  })
+//Got RED
+//   FAIL  ./code-challenges.test.js
```
Comment (SunkissedQueen) [inline code-challenges.test.js:95]: good failure

File: code-challenges.test.js [hunk 5]
```diff
@@ -15,42 +15,252 @@
...
+//tools and research: .includes checks a string against a specific string and returns true/false. It is like it iterates over the string and combines a boolean in one. .forEach could look over each value in the array and compare it to include? or .filter could return a subset of the array with the filtered values taken out?
+
+// const arrayMatch = (array, string) => {
+//    return array.filter(value => value.toLowerCase().includes(string))
+// } //since it is almost one line could I type it inline?
+
+const arrayMatch = (array, string) => array.filter(value => value.toLowerCase().includes(string))
+//the return is implicit in ES6 functions. The arrow points to the code block which is the return. I usually have difficulty with filter, but not today!
+
+// Got GREEN
```
Comment (SunkissedQueen) [inline code-challenges.test.js:135]: green!!!

File: code-challenges.test.js [hunk 6]
```diff
@@ -15,42 +15,252 @@
...
+  //if true push that index into a new array
+//tools: for loop iterates over the array, .map(), .filter(), .forEach can also iterate over an array. if statement/ternary to make a decision 
+//tools and research: .includes checks a string against a specific string and returns true/false. It is like it iterates over the string and combines a boolean in one. .forEach could look over each value in the array and compare it to include? or .filter could return a subset of the array with the filtered values taken out?
+
+// const arrayMatch = (array, string) => {
+//    return array.filter(value => value.toLowerCase().includes(string))
+// } //since it is almost one line could I type it inline?
+
+const arrayMatch = (array, string) => array.filter(value => value.toLowerCase().includes(string))
+//the return is implicit in ES6 functions. The arrow points to the code block which is the return. I usually have difficulty with filter, but not today!
```
Comment (SunkissedQueen) [inline code-challenges.test.js:133]: great pseudocode. good job with filter

File: code-challenges.test.js [hunk 7]
```diff
@@ -15,42 +15,252 @@
...
+// fullHouse
+//   ✓ takes in an array of 5 numbers and determines whether or not the array is a full house.
+
+// Test Suites: 1 passed, 1 total
+// Tests:       3 passed, 3 total
+
+// refactor attempt with ternary inline
+
+//this is crazy! completely unreadable but it is done in one line lol! deffinately more readable to store the sorted array in a variable and use that variable
+const fullHouse = (array) => (array.sort((b , a) => b - a)[0] === array.sort((b , a) => b - a)[1] && array.sort((b , a) => b - a)[0] === array.sort((b , a) => b - a)[2] && array.sort((b , a) => b - a)[3] === array.sort((b , a) => b - a)[4] || array.sort((b , a) => b - a)[0] === array.sort((b , a) => b - a)[1] && array.sort((b , a) => b - a)[2] === array.sort((b , a) => b - a)[3] && array.sort((b , a) => b - a)[2] === array.sort((b , a) => b - a)[4] ? true : false)
```
Comment (SunkissedQueen) [inline code-challenges.test.js:253]: Yes, one-liners can be helpful but I agree with you. It is more readable with variables

File: interview-questions.md [hunk 8]
```diff
@@ -7,53 +7,53 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-  Your answer:
+  Your answer: Ruby on Rails is a collection of Gems. It is a framework created in Ruby that uses the MVC framework. MVC is Model, View, Controller. Model is a layer that contains all of the business aspects of an application, the View layer contains all of the visual aspects of an application, and the Control layer contains all of the logic aspects of an application. I have used Ruby on Rails to create a small database of movies and manipulate the tables by adding new titles, the length of movies, and their genre.
 
-  Researched answer:
+  Researched answer: Ruby on Rails is a collection of gems that allow the user to create a rails application - which is server sided. It uses the MVC framework. Model, a layer that handles the logic or the business center of an application. It uses active model. View, the layer that handles what the user visually sees. View uses action view. Controller, the brains or decision maker. The Controller handless http requests and uses action pack. I have used Ruby on Rails to make an application that holds a database. The database contained a table of Movies with titles, genres, year, and length for the movie.
```
Comment (SunkissedQueen) [inline interview-questions.md:28]: marvelous

File: interview-questions.md [hunk 9]
```diff
@@ -7,53 +7,53 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 
 
 4. What is a relational database? Are there other kinds of databases?
 
-  Your answer:
+  Your answer: A relational database is a data that is stored in a table. The table has columns and rows used to organize the data. Each row is assigned a primary key that is unique to each set of data. The data in the table usually relates to each other. For example a database can contain cars. Each car could have a make, year, engine type, etc. There are many other types of databases that include cloud, commercial, object oriented.
 
-  Researched answer:
+  Researched answer: A relational databse is a database with data stored in tables. The data is then organized by columns and rows. Each row is represented by a primary key. The data in the row is related in some way. For example a database of cars could have columns named for make, year, engine type, price. Each row would contain a new car with that specific data. Ruby on Rails uses Active Record, which is a relational database.
```
Comment (SunkissedQueen) [inline interview-questions.md:36]: splendid

File: interview-questions.md [hunk 10]
```diff
@@ -7,53 +7,53 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 
 
 5. What are primary keys? Why are they important?
 
-  Your answer:
+  Your answer: A primary key is a unique identifier used in relational databases. A primary key is assigned to each row of data in the database. If the row of data is deleted, so is the primary key. A primary key that is deleted could never be used again, and is never replaced. Primary keys are important because it allows the data to be uniquely identified. It helps store, access, and manipulate specific data.
 
-  Researched answer:
+  Researched answer: A primary key is a unique identifier for database tables. In Ruby on Rails the primary keys are automatically created. If a row of data is deleted or removed from the database so is the primary key. The key is then never re-created or re-used. They are important because it is the organizational structure for that database.
```
Comment (SunkissedQueen) [inline interview-questions.md:44]: great

File: interview-questions.md [hunk 11]
```diff
@@ -7,53 +7,53 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 
 
 2. What is a gem?
 
-  Your answer:
+  Your answer: A gem in ruby is a collection of pre-built code that can be installed and used. It is like node modules for JavaScript. Ruby on Rails is a gem that allows the user to make a ruby application, 
 
-  Researched answer:
+  Researched answer: A gem is a collection of code put together that allows the user to use that code inside of their own. In Ruby gems are like adding dependencies in JavaScript. In the past I have used Rspec, a collection of code that allows the user to create tests for Ruby code.
```
Comment (SunkissedQueen) [inline interview-questions.md:20]: awesome

File: interview-questions.md [hunk 12]
```diff
@@ -7,53 +7,53 @@ Then, research the question to expand on your answer. Even if you feel you have
 
 1. What is a hash in Ruby?
 
-  Your answer:
+  Your answer: A hash in Ruby is a dictionary like set of data. It is similar to data stored in an object in that there is a 'key' and a 'value'. The key is a symbol, and the values in a hash can be any other data. Since Ruby uses duck type, a Hash is similar to a list of data. It is enumarable, meaning methods that work on enumerable data work on Hashes. I have used map, forEach, and regex in functions on hashes.
 
-  Researched answer:
+  Researched answer: Ruby Hashes are a dictionary-like set of keys and values. Since hashes are associative arrays, they use any data type instead of numbers as an index. They keys in a hash are symbols. Hashes are considered enumerable because Ruby uses duck type. Duck type is when something is similar to another thing, the same methods should work on both. Since hashes are like lists, methods that work on lists like map, and forEach would work.
```
Comment (SunkissedQueen) [inline interview-questions.md:12]: nice

File: ruby-challenges.rb [hunk 13]
```diff
@@ -10,6 +10,26 @@
 letter_t = 't'
 # Expected output: ['tea', 'water', 'soda water']
 
+# create a method match_word that takes in an array and a string(letter)
+# the method then checks the array to contain the letter
+# the output is then a subset of the array with only the words containing that letter
+# tools and research: .include - returns true/false; .map iterates over the array but always returns the same array, filter returns a subset
+
+# takes in two arguments
+def match_word(array , string)
```
Comment (SunkissedQueen) [inline ruby-challenges.rb:19]: Why did you use an empty array to store the array created by `filter`?

File: ruby-challenges.rb [hunk 14]
```diff
@@ -19,17 +39,91 @@
...
 
+# method sum_up 
+# it takes in an array of integers
+# expected output is the sum of all of those integers as an integer
+# research: .sum apparently just adds all integers in an array
+
+# define the method, takes in an array 
+def sum_up array 
+    # .sum is a built in method that adds all integers.
+    array.sum
```
Comment (SunkissedQueen) [inline ruby-challenges.rb:50]: sum is winning

File: ruby-challenges.rb [hunk 15]
```diff
@@ -19,17 +39,91 @@
...
+
+end
+
+p sum_up nums_array1 # output-> 76
+p sum_up nums_array2 # output-> 100
 
 
 # --------------------3a) Create a class called Bike that is initialized with a model, wheels, and current_speed. The default number of wheels is 2. The current_speed should start at 0. Create a bike_info method that returns a sentence with all the data from the bike object.
+# create a class of Bike
```
Comment (SunkissedQueen) [inline ruby-challenges.rb:59]: beautiful pseudocode and class

**Review by SunkissedQueen — APPROVED** [review]
Stephen, you are such a thoughtful developer. Maintain your performance for the next assessment. With the interview questions and your coding, you will be well on your way to sharing your talent with any web development platform. Great work! Next assessment after you are gotten the code block to work, go through and see if there are any redundant methods. Since you are on top of your vocabulary, suggest using the research to find something that you didn't know about the tool.

File: ruby-challenges.rb (hunk 13 above)
Comment (SOMarrah) [reply ruby-challenges.rb:19]: That is a good question, the filter should return the values I put in the conditional -> i forgot to not use if statements in the filter! `def match_word(array , string)
    array.filter do |value|
        value.include? string
        end
    end
end`

Maybe the update would look something like that? Havent tested it, just thought about it.

**Pull request merged**